documentation/design/proposals/authn_authz_for_1.0.adoc

```diff
@@ -0,0 +1,273 @@
+Admin, or alternatively the Service Admin will be able to configure one or more Access Control Services centrally. Selectors will automatically associate an Access Control Service with the Tenants.
+
+For the getting started use case, there will be a implict in-built default which will apply until the first Access Control
+Service is defined by the Service Admin.  The in-built default will provide ANONYMOUS access with an allow all policy.
+
+== Design overview
+
+=== Compoenents
+
+==== Access Control Server
+
+An Access Control Server instance will provide identity management and authorization services to all the tenants of an instance of shared infrastructure.  The Access Control Server will be a stateless component that will be scaleable for
+availability and performance.
+
+The Access Control Server will be created by EnMasse Operator according to configuration within the MessagingInfra object.
+
+The Access Control Server will act on the `AccessControlService`, `MessagingUser`, `MessagingRole`, and `MessagingRoleBinding` objects that exist in the scope defined by the `MessagingInfra` (i.e. its namespace selector).  It
+will act as the controller for all of these objects and be responsible for maintaintence of the status sections.  In addition it will watch `ServiceAccounts` thus enabling the service account authentication use-case. 
+
+The Access Control Server will present an AMQP SASL endpoint, implementing the `authServicePlugin` interface, per
+messaging tenant.  It will be the responsibility of the EnMasse Operator to configure a router listener _per tenant_ each
+with its own authServicePlugin configuration referencing the correct endpoint.  TLS SNI will allow the Access Control Server
+to work out which `AccessControlService` to apply when processing each authentication event.
+
+The algorithm for processing each authentication event:
+
+1. Router connects to the AMQP SASL endpoint associated with the router listener.
+1. Access Control Server uses the SNI information to resolve the applicable `AccessControlService`
+1. Access Control Service uses the configuration of the `AccessControlService` determine the avaiable SASL mechanism(s).
+1. SASL-MECHANISMS performative flows the Client.
+1. Client selects mechanism
+1. SASL-INIT flows to the Access Control Server
+1. SASL-CHALLENGE / SASL RESPONSE as necessary
+1. Access Control Server calls out to the external authentication service / uses `MessagingUser`/`ServiceAccount` (as appropiate) to determine if the authentication is successful.
+1. SASL-OUTCOME flows to the Client.
+1. Access Control Server uses the `MessagingRole`, and `MessagingRoleBinding` to calculate the `ADDRESS-AUTHZ`.
+1. AMQP OPEN flows to the Router carrying the authorisation policy.
+
+When the messaging clients authenticate, the SASL username passed by the application must be qualified with a domain suffix (`@DOMAIN`). This allows the Access Control Server to disambiguate identities between providers. 
```

Comment:
Couldn't we also use the 'realm'/'saslInitHost' that is provided by the router auth plugin?

Reply:
This was may first idea, unfortunately it doesn't work.  `saslInit` is too late.  The server needs to provide the SASL mechanisms response before the client sends saslInit. In order to do this, the server needs to know what mechanisms to present. To do this, is needs to be able to resolve the authentication configuration.  I think using SNI is the only option.

Reply:
Given that its possible to omit it for service accounts and messaging users , I think that is sufficient.